Add self-diffusivity calculation

xhgchen opened the pull request:
Fixes #7 

Changes made in this Pull Request:
 - Add self-diffusivity calculation method `sd()` in class `VelocityAutocorr`

xhgchen commented:
Definitely want to discuss what is best for testing this, perhaps during our meeting on Thursday. I did have some ideas in my proposal, though none of them are as straightforward to put into practice as the tests written thus far.

hmacdope commented:
The step traj should have an analytically solvable integral, though what it is I am not sure.

xhgchen commented:
Is it really easily solvable analytically? I'm thinking about trying for a different numerical integration method e.g. `scipy.integrate.quad` for the tests.

hmacdope commented:
Try making a few notebooks to play around in and we can go over it on when we meet soon. :+1:

orionarcher commented:
@reviewnb

xhgchen commented:
> @ReviewNB
> 
> This _should_ make it easier to integrate notebooks into review/version control if you want to do that.

Thank you, I'll look into it!

xhgchen commented:
Added tests for both the simple windowed calculation and the FFT calculation but not sure if both are necessary - please let me know your thoughts. Using `scipy.integrate.simpson` for the tests to start and changed `NSTEP` from 5000 to 5001 to work with that method.

xhgchen commented:
We could probably do without the non-FFT tests for self-diffusivity, since we test that already in the VACF tests. Am I good to remove them?

hmacdope commented:
I would check the full code path, testing is cheap but having undetected issues is expensive.

xhgchen commented:
Made the applicable changes @orionarcher requested in #25 here, namely changing `_parse_dim_type()` to a static method and removing the unnecessary type casting.

hmacdope commented on transport_analysis/tests/test_velocityautocorr.py:
```diff
@@ -286,10 +427,146 @@ def test_fft_start_stop_step_all_dims(
+            )
+            / tdim_factor
+        )
+        # 7705160166.66 (exp) agrees with 7705162888.88 (act) to 6 sig figs
```
Is this really the self diffusivity? Seems very high. Try with the equivalent trajectory in positions space with the existing MSD module from MDAnalysis and see what you get.

xhgchen replied:
Tests have been added (though they are very rough). Will try to make the separate class next!

hmacdope reviewed (changes requested):
I think we need to validate against existing MSD code.

hmacdope commented on transport_analysis/tests/test_velocityautocorr.py:
```diff
@@ -286,10 +427,146 @@ def test_fft_start_stop_step_all_dims(
-        assert_almost_equal(v_simple.results.timeseries, poly, decimal=3)
+        assert_almost_equal(v_fft.results.timeseries, poly, decimal=3)
+
+    @pytest.mark.parametrize(
```
You can separate all the tests that use this fixture into a separate class and then make the parameterization over  the class rather than repeat

xhgchen replied:
Those tests would include the old ones for VACF too, right?

hmacdope commented on transport_analysis/velocityautocorr.py:
```diff
@@ -262,3 +261,68 @@ def plot_vacf(self, start=0, stop=0, step=1):
             self.results.timeseries[start:stop:step],
         )
+
+    def sd(self, start=0, stop=0, step=1):
```
This is a green Kubo integral of the VACF so we should use the proper name.

hmacdope replied:
Also will need citations.

xhgchen replied:
Does `sd_gk()` sound good? I feel that `self-diffusivity_greenkubo()` is too long. I thought we were doing citations after we finish the bulk of the implementations? Or would it be better to do them now?

hmacdope replied:
`sd_gk` sounds good to me. Either is fine on citation, up to you, just make a note or issue.

xhgchen replied:
Left a note in Issue #20.

hmacdope commented on transport_analysis/velocityautocorr.py:
```diff
@@ -262,3 +261,68 @@ def plot_vacf(self, start=0, stop=0, step=1):
+        `numpy.float64`
+            The calculated self-diffusivity value for the analysis.
+        """
+        stop = self.n_frames if stop == 0 else stop
```
Should raise an exception if .run hasn't already been called with some kind of cache variable.

xhgchen replied:
I'm not too sure where to start here. Do you think you'd be able to explain a bit more, or would you be good with setting up a meeting to go over this?

hmacdope replied:
Yep happy to meet. 

Something like the following should work. 


```python

def __init__(self,)
    self._run_called = False
   ...

   def _conclude(self):
   self._run_called = True
   ...

   def sd_gk(self):
    if not self._run_called:
        raise RunTimeError("blah blah")
```

orionarcher replied:
Or `self._has_run`?

Something like, `"VelocityAutocorrelation.run() must be called before 'sd_gk'"`

hmacdope commented on transport_analysis/velocityautocorr.py:
```diff
@@ -262,3 +261,68 @@ def plot_vacf(self, start=0, stop=0, step=1):
+            / self.dim_fac
+        )
+
+    def sd_odd(self, start=0, stop=0, step=1):
```
We will probably remove this IMHO as trapeziod or Gquad is the standard. Can keep it around for testing but I would make it a private method with a `_` prefix

xhgchen replied:
I thought Simpson's rule was better when appropriate? Here's a link to my source: https://math.dartmouth.edu/~m3cod/klbookLectures/406unit/trap.pdf

Is keeping the standard more important than having an option for better accuracy?

hmacdope replied:
Need for an odd number of samples is a bit of a pain.

xhgchen replied:
I feel having the option available is better than not providing it, but I am no expert. Is it really preferable to remove it rather than just make it clear that trapezoid is the standard/default?

hmacdope replied:
In general you want to aim for the lowest possible API surface so that people don't get confused and do the wrong thing. 

The idea being you should plan for users to get frustrated as soon as the first thing goes wrong, meaning you should take away the ability for them to annoy themselves if that makes sense.

xhgchen replied:
If users do the wrong thing (use the method with an even number of samples), `scipy.integrate.simpson` actually handles it well by default using the equations in Cartwright's paper for the last interval. On second thought, considering that it is pretty much the same as trapezoid but better, what are your thoughts on using it as the default instead?

Source: https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.simpson.html

hmacdope reviewed (changes requested):
Looking great, we should also add something to plot the running integral of the ACF as is done in.  Fig 12 of https://livecomsjournal.org/index.php/livecoms/article/view/v1i1e6324/937. The right function for this is probably `scipy.integrate.cumtrapz`.

orionarcher commented on transport_analysis/tests/test_velocityautocorr.py:
```diff
@@ -42,14 +49,29 @@ def vacf(ag):
 
+# Position trajectory corresponding to unit velocity trajectory
+@pytest.fixture(scope="module")
+def step_vtraj_pos(NSTEP):
+    x = np.arange(NSTEP).astype(np.float64)
```
Can we make the name here a bit more descriptive? Maybe `unit_velocity_traj`

xhgchen replied:
Does this matter enough to be worth the effort? It's used in a lot of places, so the change would be a little messy.

orionarcher commented on transport_analysis/velocityautocorr.py:
```diff
@@ -262,3 +266,78 @@ def plot_vacf(self, start=0, stop=0, step=1):
             self.results.timeseries[start:stop:step],
         )
+
+    def sd_gk(self, start=0, stop=0, step=1):
```
Another nitpick, but generally I'm not in favor of suber abbeviated naming. `sd_gk` means something to you now, as a developer, but it probably won't mean much to a new user. Maybe `self_diffusivity_gk` or `self_diffusivity_green_kubo` or `diffusivity_green_kubo`?

xhgchen replied:
Good point, I'll go with `self_diffusivity_gk` then.

hmacdope reviewed (approved):
Great work @xhgchen! Addressed all my comments. I would say merge and lets kick some more goals. :100:
